# Docker-image-of-DNBelabC4_scRNA-analysis

## 1. Introduction

 The original version of DNBelab C Series Single-Cell RNA Analysis pipeline was released by MGI-tech-bioinformatics (https://github.com/MGI-tech-bioinformatics/DNBelab_C_Series_scRNA-analysis-software). This is an updated docker version. The image has been tested by MGI-EU team on 2021-08-12.

### 1.1 Purpose

- An open source and flexible pipeline to analyze the raw fastq files generated by  DNBelab C Series scRNA-seq platform
- Package the pipeline's working environment into a docker container image.
- Adjusted the internal file structure of the image so that it can be Singularity-ized and  run without root access.

### 1.2 Workflow

![img](https://github.com/MGI-tech-bioinformatics/DNBelab_C_Series_scRNA-analysis-software/raw/master/doc/fig/workflow.jpg)



## 2. Overview

### 2.1 Demo

#### 2.1.1 Demo data

Here we provide 2 demos (single species and double species) run in singularity. All the output results are provided in the example file.

- Single species

We provide [100MB demo human scRNA sequencing data](https://bgitech-my.sharepoint.com/personal/huangshunkai_genomics_cn/_layouts/15/onedrive.aspx?id=%2Fpersonal%2Fhuangshunkai%5Fgenomics%5Fcn%2FDocuments%2FDNBelab%20C4%2FscRNA%2EDemo%2Fhuman%2Edemo&originalPath=aHR0cHM6Ly9iZ2l0ZWNoLW15LnNoYXJlcG9pbnQuY29tLzpmOi9nL3BlcnNvbmFsL2h1YW5nc2h1bmthaV9nZW5vbWljc19jbi9FaGJjSXA1aUtlMUtzcTFQZi03djB5SUJrV05KdkgwOEpYODlyZGxQT1UxS1ZRP3J0aW1lPWdkZWVqQ1JlMlVn) for testing. And for larger dataset, we provide [36GB PBMC sample fastq data](https://bgitech-my.sharepoint.com/personal/huangshunkai_genomics_cn/_layouts/15/onedrive.aspx?id=%2Fpersonal%2Fhuangshunkai%5Fgenomics%5Fcn%2FDocuments%2FDNBelab%20C4%2FscRNA%2EDemo%2Fhuman%2Edemo&originalPath=aHR0cHM6Ly9iZ2l0ZWNoLW15LnNoYXJlcG9pbnQuY29tLzpmOi9nL3BlcnNvbmFsL2h1YW5nc2h1bmthaV9nZW5vbWljc19jbi9FaGJjSXA1aUtlMUtzcTFQZi03djB5SUJrV05KdkgwOEpYODlyZGxQT1UxS1ZRP3J0aW1lPWdkZWVqQ1JlMlVn).

-  Double species

We provide [135 MB double species(GRCh38 & mm10) demo data](https://bgitech-my.sharepoint.com/personal/huangshunkai_genomics_cn/_layouts/15/onedrive.aspx?id=%2Fpersonal%2Fhuangshunkai%5Fgenomics%5Fcn%2FDocuments%2FDNBelab%20C4%2FscRNA%2EDemo%2Fmix%2Edemo&originalPath=aHR0cHM6Ly9iZ2l0ZWNoLW15LnNoYXJlcG9pbnQuY29tLzpmOi9nL3BlcnNvbmFsL2h1YW5nc2h1bmthaV9nZW5vbWljc19jbi9FbmVFMk84VU9aeENqNVM3RlRTZmRROEJHaHRrMkt4dVNicUNsSEo3bC1uS0x3P3J0aW1lPWFKUmJLaVZlMlVn) for testing. We also provide [52 GB scRNA sequencing data on mixed sample](https://ftp.cngb.org/pub/CNSA/data2/CNP0000906/CNS0196715/CNX0144387/CNR0177382/) (GRCh38 & mm10).

#### 2.1.2 Demo execution

For single species,

```shell
#Step1. Check config.json file
$cat config.json
{
    "main.fastq1": "/DNBelab_C4/rawfq/Demo1.human.fq.1.gz",                 
    "main.fastq2": "/DNBelab_C4/rawfq/Demo1.human.fq.2.gz",                 
    "main.ID": "Demo_single",                                        
    "main.expectCell":"500",                                          
    "main.umilow": "1000",                                           
    "main.species":"GRCh38",                                    
    "main.original":"cell lines",                            
    "main.SampleTime":"2020-06-25",                         
    "main.ExperimentalTime":"2020-06-25"                    
   }
#Step2. Run pipeline
/app/singularity-3-2/bin/singularity exec \
--bind /DNBelab_V1_demo_test/database/GRCh38:/database \
--bind /DNBelab_V1_demo_test/data:/rawfq \
--bind /DNBelab_V1_demo_test/output:/result \
dnbelab_c4.sif /bin/bash /DNBelab_C4/bin/mgiRun.sh
```

For double species,

```shell
#Step1.Check config.json file
$cat config.json
{
    "main.fastq1": "/DNBelab_C4/rawfq/Demo.double.fq.1.gz",
    "main.fastq2": "/DNBelab_C4/rawfq/Demo.double.fq.2.gz",
    "main.ID": "Demo_double",
    "main.umilow": "1000",
    "main.expectCell":"500",
    "main.species":"GRCh38_mm10",
    "main.original":"cell lines",
    "main.SampleTime":"2021-06-25",
    "main.ExperimentalTime":"2021-06-25"
   }
#Step2. Run pipeline
/app/singularity-3-2/bin/singularity exec \
--bind /DNBelab_V1_demo_test/database/GRCh38_mm10:/database \
--bind /DNBelab_V1_demo_test/data/double_speices:/rawfq \
--bind /DNBelab_V1_demo_test/double_demo/output:/result \
/DNBelab_V1_demo_test/dnbelab_c4.sif /bin/bash /DNBelab_C4/bin/mgiRun_double.sh
```

#### 2.1.3 Demo output

```shell
# After all analysis processes ending, you will get these files below(both single and double species):
cd /DNBelab_V1_demo_test/double_demo/output && ls
outs/  report/  temp/  /symbol workflowtime.log

ls out
cell_barcodes.txt  cluster.h5ad  count_mtx.tsv.gz  final.bam

ls report
alignment_report.csv  annotated_report.csv  cell_report.csv  cluster.csv  cutoff.csv  iDrop_demo.html  marker.csv  RNA_counts.pdf  sample.csv  sequencing_report.csv  vln.csv

ls symbol
# In single_Species result,there are some follow files will be generated:

makedir_sigh.txt parseFastq_sigh.txt fastq2bam_sigh.txt sortBam_sigh.txt cellCount_sigh.txt cellCalling_sigh.txt countMatrix_sigh.txt report_sigh.txt
```

Html report

- single species

![image](https://user-images.githubusercontent.com/79911830/129350861-60f2ca1a-941d-46da-b3be-18b6e5aeaa2c.png)

- Double species
![image](https://user-images.githubusercontent.com/79911830/129350982-fe70d961-a2e5-4921-93cd-f26681e31be5.png)

### 2.2 Latest update

Detailed updates from [original repository](https://github.com/MGI-tech-bioinformatics/DNBelab_C_Series_scRNA-analysis-software)

1. Add the function of double species data analyzing to docker image.
2. Adjusted the internal file structure of the image so that it can be Singularity-ized and  run without root access.

## 3. **Hardware/Software requirements**

 ### 3.1 Hardware 

- x86-64 compatible processors
- 64 bit Linux
- At least 36GB of RAM and 10 CPUs

### 3.2 Software

- Docker ≥ 20.10.6

- Sudo permission

  or

- singularity ≥  3.2.0-1

### 3.3 Supplementary

The following requirements has been deployed within the image, you can execute them in the container. 

- [java](https://www.oracle.com/java/)
- [Cromwell](https://github.com/broadinstitute/cromwell/releases)
- [R(3.5+)](https://www.r-project.org/) #with following R packages installed 
  - ggplot2
  - getopt
  - data.table
  - cowplot
  - DropletUtils(1.6.1+)
- [python3 (3.6+)](https://www.python.org/downloads/) # with following Python3 packages installed 
  - numpy
  - pandas
  - python-igraph
  - louvain
  - scanpy(1.4.3+)
  - jinja2(2.10.3+)
- [PISA](https://github.com/shiquan/PISA/wiki)
- [sambamba](https://lomereiter.github.io/sambamba/)
- [STAR](https://github.com/alexdobin/STAR)


## 4. Database and files requirements

### 4.1 Database

A database containing the STAR index for reference genome and its annotation files is needed. Here we provide [3 read-made datasets](https://ftp.cngb.org/pub/CNSA/data2/CNP0000906/)。

- Human (GRCh38)
- Mouse (GRCm38)
- Mixed Database(GRCh38&GRCm38)

You can also build the database by yourself.

```shell
# Creat the database directory
mkdir database/example_database && cd database/example_database
mkdir fasta && mkdir gtf && mkdir star_index 
# Copy or download the prepared fastq file and gene gtf file to the corresponding directory
# Build STAR index
PATH/TO/STAR --runThreadN 8 --runMode genomeGenerate --genomeDir star_index --genomeFastaFiles ./fasta/example_genome.fa --sjdbGTFfile ./gtf/example_genes.gtf
```

Note: 

- It takes about 1 hour to build the index file for a 3G genome;
- The STAR version for index-building needs to be consistent with the  STAR version used for mapping. In this docker image we use V2.7.3 STAR. It is recommended to download the executable STAR file we provided, thus ensuring the consistency of STAR version.

### 4.2 Config files

#### 4.2.1 Examples

An config.json file containing all the input parameters is needed for running the pipeline. Mandatory parameters must be specified in the documentation, and the pipleline will use default values if the optional parameters are not specified in the config file. 

```shell
#Example
cat config.json
   {
    "main.ID": "Demo_single",  
    "main.fastq1": "/DNBelab_C4/rawfq/demo_1.fq.gz",                 
    "main.fastq2": "/DNBelab_C4/rawfq/demo_2.fq.gz",                                                    
    "main.forceCell": "0",                                          
    "main.umilow": "1000",                                           
    "main.species":"GRCh38",                                    
    "main.original":"cell lines",                            
    "main.SampleTime":"2020-06-25",                         
    "main.ExperimentalTime":"2020-06-25"                    
   }
```

#### 4.2.2 Parameters

The following table explains the parameters of the config file.

| Parameter             | Type      | Description                                                  |
| --------------------- | --------- | ------------------------------------------------------------ |
| main.ID               | String    | MANDATORY. Sample id.                                        |
| main.fastq1           | Fastq     | MANDATORY. Read 1 in fastq format. Can be gzipped. **Do not change the file path(/DNBelab_C4/rawfq/XXX.fq)**, only need to change the name of input fastq files. Fastqs from different lanes can be separated with comma. For example, "L01_read_1.fq.gz, L02_read_1.fq.gz,...". |
| main.fastq2           | Fastq     | MANDATORY. Read 2 in fastq format. Can be gzipped. **Do not change the file path(/DNBelab_C4/rawfq/XXX.fq)**, only need to change the name of input fastq files. Fastqs from different lanes can be separated with comma. For example, "L01_read_2.fq.gz, L02_read_2.fq.gz,...". |
| main.forceCell        | Integer   | Optional, default: 0. Specified number of cells.             |
| main.expectCell       | Integer   | optional, default: 0. Expected cell number.                  |
| main.umiCount         | Integer   | Optional, default: 0. Threshold of barcode's umi, anything above that would be defined as contribution from vaid cells. |
| main.umilow           | Integer   | Optional, default: 50. Expected UMIs count per cell threshold. |
| main.chrom            | File Path | Optional, default: `config/species_binding.txt` directory. Chromosome and species correspondence files in the mixed species process. And Single species process don't need this file. |
| main.species          | String    | Optional, default: Null. Species.                            |
| main.original         | String    | Optional, default: Null. original.                           |
| main.SampleTime       | String    | Optional, default: Null. Sample time.                        |
| main.ExperimentalTime | String    | Optional, default: Null. Experimental time.                  |

### 4.3 Raw Fastq

Raw fastq files containing scRNA sequence from both DNBelabC platform and 10X Genomics platform can be taken as input fastq.

## 5. Installation(Root access needed)

### 5.1 Docker 

```shell
#1.Install docker follow the official tutorial: https://www.docker.com/
#2. Pull the images from docker hub: 
docker pull dingrp/huangshunkai_dnbelab_c4:latest
```

### 5.2 Singularity

```shell
#1. Install singularity follow the offical tutorial:https://sylabs.io/guides/3.5/user-guide/quick_start.html#
#2. Pull & build Singularity image
singularity build DNBelabC4.sif docker://dingrp/huangshunkai_dnbelab_c4:latest
```

## 6. Usage

### 6.1 Run with docker

```shell
Preparation: 

1. Confirm the database and raw fastq files being prepared as described in 3.1 and 3.3;
2. Confirm the cong.json file being setted according to 3.2.

Running

1. Please set the following variables on your machine:
   (a) $DB_LOCAL: Directory on your local machine that has the database files. Make sure that the directory must contains two subdirectories, "gtf" and "star_index". The gene annotation file named "genes.gtf" must be included under "gtf"; the genome index file for STAR under the "star_index". If you build the database youself, make sure the format of the directory path is correct.
   (b) $DATA_LOCAL: Directory on your local machine that has the fastq fil and "config.json" file.
      "config.json" must follow the format descripted in 3.2,
      and the *PATH* in "config.json" must be absolute dicrtory of $DATA_LOCAL.
   (c) $RESULT_LOCAL: directory for output results.
   (d) $scRNANAME: The name of the container that you want to assign to.
  
2. Run the command:
   
   mgi single species data:
   docker run -d -P \
   --name $scRNANAME \
   -v $DB_LOCAL:/DNBelab_C4/database \
   -v $DATA_LOCAL:/DNBelab_C4/rawfq \
   -v $RESULT_LOCAL:/DNBelab_C4/result \
   dingrp/huangshunkai_dnbelab_c4:latest \
   /bin/bash \
   /DNBelab_C4/bin/mgiRun.sh
   
   mgi double species data:
   docker run -d -P \
   --name $scRNANAME \
   -v $DB_LOCAL:/DNBelab_C4/database \
   -v $DATA_LOCAL:/DNBelab_C4/rawfq \
   -v $RESULT_LOCAL:/DNBelab_C4/result \
   dingrp/huangshunkai_dnbelab_c4:latest \
   /bin/bash \
   /DNBelab_C4/bin/mgiRun_double.sh
   
   10x single species data:
   docker run -d -P \
   --name $scRNANAME \
   -v $DB_LOCAL:/DNBelab_C4/database \
   -v $DATA_LOCAL:/DNBelab_C4/rawfq \
   -v $RESULT_LOCAL:/DNBelab_C4/result \
   dingrp/huangshunkai_dnbelab_c4:latest \
   /bin/bash \
   /DNBelab_C4/bin/10xRun.sh
  
   

3. After satisfactory result was generated:
   docker rm $scRNANAME
```

### 6.2 Run with singularity

```shell
Preparation: 

1. Confirm the database and raw fastq files being prepared as described in 3.1 and 3.3;
2. Confirm the cong.json file being setted according to 3.2.

Running

1. Please set the following variables on your machine:
   (a) $DB_LOCAL: Directory on your local machine that has the database files. Make sure that the directory must contains two subdirectories, "gtf" and "star_index". The gene annotation file named "genes.gtf" must be included under "gtf"; the genome index file for STAR under the "star_index". If you build the database youself, make sure the format of the directory path is correct.
   (b) $DATA_LOCAL: Directory on your local machine that has the fastq fil and "config.json" file.
      "config.json" must follow the format descripted in 3.2,
      and the *PATH* in "config.json" must be absolute dicrtory of $DATA_LOCAL.
   (c) $RESULT_LOCAL: directory for output results.
   (d) $scRNANAME: The name of the container that you want to assign to.

2. Run the command

mgi single species data:
/Path/to/singularity exec \
--bind $DB_LOCAL:/database \
--bind /$DATA_LOCAL:/rawfq \
--bind $RESULT_LOCAL:/result \
/Path/to/DNBelabC4.sif /bin/bash /DNBelab_C4/bin/mgiRun.sh

mgi double species data
/Path/to/singularity exec \
--bind $DB_LOCAL:/database \
--bind /$DATA_LOCAL:/rawfq \
--bind $RESULT_LOCAL:/result \
/Path/to/DNBelabC4.sif /bin/bash /DNBelab_C4/bin/mgiRun_double.sh

10x single species data:
/Path/to/singularity exec \
--bind $DB_LOCAL:/database \
--bind /$DATA_LOCAL:/rawfq \
--bind $RESULT_LOCAL:/result \
/Path/to/DNBelabC4.sif /bin/bash /DNBelab_C4/bin/10xRun.sh
```

### 6.3 Output Results

After the running is completed,  results are output in the $RESULT_LOCAL in your locat machine.

```shell
# After all analysis processes ending, you will get these files below:
cd result && ls
outs/  report/  temp/  /symbol workflowtime.log

ls out
cell_barcodes.txt  cluster.h5ad  count_mtx.tsv.gz  final.bam

ls report
alignment_report.csv  annotated_report.csv  cell_report.csv  cluster.csv  cutoff.csv  iDrop_demo.html  marker.csv  RNA_counts.pdf  sample.csv  sequencing_report.csv  vln.csv

ls symbol
# In single_Species result,there are some follow files will be generated:

makedir_sigh.txt parseFastq_sigh.txt fastq2bam_sigh.txt sortBam_sigh.txt cellCount_sigh.txt cellCalling_sigh.txt countMatrix_sigh.txt report_sigh.txt
```

- **Cell×Gene expression matrix**

  *$RESULT_LOCAL/outs/count_mtx.tsv.gz*.  The mtx file can be loaded into the general scRNA data analyzing tools like Seurat and scanpy, for further analysis.

- **Html Analyzing report**

  *$RESULT_LOCAL/report/iDrop_Demo_single.html.* This is a detailed analyzing report for the pipline. It reveals the reuslts of basecalling, as well as the distribution of UMI counts, Gene counts in the identified cells. Cell clustering results are also exibited in the html report, including highly expressed genes for each cluster.

- **Resulting report for each step**

  *$RESULT_LOCAL/report/*  This fold contains the running reports for each Intermediate steps, including fastq QC, alignment, annotation cell calling and clustering.

- **Intermediate results files for each step**

  *$RESULT_LOCAL/temp* 

## 6. Updates

Detailed updates from [original repository](https://github.com/MGI-tech-bioinformatics/DNBelab_C_Series_scRNA-analysis-software)

1. Add the function of double species data analyzing to docker image.
2. Adjusted the internal file structure of the image so that it can be Singularity-ized and  run without root access.

## FAQ

1. Why you only provide the docker version here?

   This pipeline involves multiple softwares and packages, which may take  a lot of efforts to install and deploy. Docker  enables the developer integrating all the softwares needed in a container, users only need to download the image from the repository and they can run it directly.

   You can also find the non-docker version from the [original repository](https://github.com/MGI-tech-bioinformatics/DNBelab_C_Series_scRNA-analysis-software).

2. Why do I get failures when I run demo samples?

   The small size of the demo sample may failed if you use the default value for main.expectCell and main.forceCell at same time. That's because the dataset is too small to be judged as complete cells by the pipeline. The extremely low number of cells being called leads to the failure in the following steps (Cell clustering). In this case we suggest you to set the main.expectCell or main.forceCell manually.

3. Why the inflection point is sometimes inaccurate on the total count curve?

   You can specify "main.umilow" in the configure file like "main.umilow": "1000". "main.umilow" is a numeric scalar specifying the lower bound on the total UMI count, at or below which all barcodes are assumed to correspond to empty droplets, default 50.

## Reference

https://github.com/MGI-tech-bioinformatics/DNBelab_C_Series_scRNA-analysis-software

## Contribution

- Renpeng Ding, MGI-Latvia

- Shunkai Huang, MGI
